proto/config/compound_devconfig.proto

```diff
@@ -15,8 +15,11 @@ import "profile/local_profile.proto";
 // needed for the LOC case, when a single "/compound-config" endpoint can be
 // used.
 message CompoundEdgeDevConfig {
-   org.lfedge.eve.auth.AuthContainer protected_config = 1;
-   org.lfedge.eve.profile.LocalDevCmd dev_cmd = 2;
-   org.lfedge.eve.profile.LocalAppCmdList app_cmd_list = 3;
-   org.lfedge.eve.profile.RadioConfig radio_config = 4;
+   // Sequence number of compound config, used to guarantee all commands
+   // are processed. See the pair field `ack_seq` of the `ConfigRequest`.
```

Comment:
I don't understand how this will be used to ensure that all commands are processed.
What are the assumptions about when the device will increment the number in the request e.g., will it do that after it has timed out connecting to the LOC? After a reboot?

A sequence number (or a random nonce) can be used to match a request with a response, but if the response is to a POST then the request (POST payload) and response to POST are naturally associated.
So not clear what problem you are trying to solve.

Reply:
The logic is similar to the LPS 'timestamp' and 'last_cmd_timestamp' which goes along with each command or info request (see the local_profile.proto). The word timestamp in case of LPS looks clumsy to me, and 'seq/ack_seq' is the default term for the acknowledgement mechanism.

I'm trying to achieve acknowledgement of processed commands, so LOC knows what has been processed, what not. Probably the statement 'guarantee all commands are processed' is incorrect here and needs re-wording. I will change this.

Reply:
Note that the reason we called the field `timestamp` was just to make it less bug prone for the developers of LPS (which is not us in that case). For correct acknowledgement of processed commands, it is required that two successive but distinct requests made will have different timestamps (aka seq. numbers) attached (mentioned [here](https://github.com/lf-edge/eve-api/blob/main/proto/profile/local_profile.proto#L113-L121)). This needs to be ensured also across LPS reboots. Simple non-persistent counter does not satisfy this requirement. Storing e.g. unix timestamp into `seq` will work. So we decided to call it `timestamp` to simply give hint to LPS developers how to develop this.
In case of LOC you are in control of its implementation so you can chose better terminology.

Reply:
Thanks @milan-zededa for the clarification!

Reply:
just to be clear: I clearly understand why you chose the timestamp and I did not want to raise any hot discussion with my "clumsy terminology" statement (although I don't take it back) :) I just  don't like the timestamp approach, because this does not guarantee you the persistence - clocks can drift in the past as well (simple scenario: no access to the internet for weeks, all good, command is issued, LPS vm reboots, connection to the internet appeared, LPS syncs time with NTP, time goes to the past, boom). Yes, I agree, much better than counter, but I don't like this "kinda works" things.

BTW, the LOC can also be rebooted. But. The problem can be easily fixed by the LOC implementation (as you said): if `ack_seq` (comes from EVE) is detected > than `seq`, then LOC can update the `seq` and set it to the max. 

Anyway. Thanks for the clarification once again.

Reply:
> The logic is similar to the LPS 'timestamp' and 'last_cmd_timestamp' which goes along with each command or info request (see the local_profile.proto). The word timestamp in case of LPS looks clumsy to me, and 'seq/ack_seq' is the default term for the acknowledgement mechanism.

sequences numbers tend to imply monotonically increasing (without gaps) numbers and potentially the need to handle wraparound from 2^N-1 to zero.
timestamp is a somewhat predictable thing.
A nonce would be a random thing which can't be guessed.
Not clear what the semantics are other than the LOC echoing the most recently received number. Can/should the LOC reject "old" sequence numbers? The choice of name would make me think that is part of the semantics. But that implies that EVE has to remember the largest sequence number it has ever used.

Reply:
And the suggestion to have the LOC not persist the largest number it has used but instead learn it from EVE's seq_ack sounds like all the semantics you get is that of a non-random nonce; you effectively can only check for equal since greater/less doesn't mean anything if you adjust the sequence number based on the seq_ack.

Thus you have a timestamp from some form of virtual clock and nothing more AFAICT. So might as well use a timestamp from the wall clock.,

If you really need and want to check for greater than, then you need something a lot more robust than what you are describing.

Reply:
In the current implementation, I don't see where "old" sequence numbers can appear and cause any problems. There are two possible scenarios where sequence numbers are out of sync:

a) EVE has been restarted. This case is dealt with in permanent files (already implemented for LPS) where all pending commands are stored with their respective timestamps/sequence numbers. EVE sends `ack_seq` equal to 0, so LOC does nothing because its `seq` number is higher. The next configuration response (sent from LOC) with `seq` number will be processed by the default `ProcessReceived*Commands` functions (all pending commands have been read from file), `ack_seq` number will be set to `seq` .

b) LOC is restarted. EVE requests a configuration with `ack_seq` greater than `seq` LOC number (which is 0 after restart). Since this is the first configuration request since the LOC restart, two scenarios are possible:
b.1) No commands were issued before the first configuration request. The LOC updates its `seq` number with the number it received from EVE, so that all future commands and responses will be tagged with the same `seq` number.
b.2) There were commands sent right after the LOC restarted and before EVE requested its first configuration. These are new commands, so they must be executed on EVE. As soon as the LOC receives the first configuration request with `ack_seq` > `seq`, the LOC sets its own `seq` = `ack_seq` + 1 and updates all issued commands accordingly. In this case, EVE always gets a `seq` number greater than it is currently processing in order to start processing new task commands.

The only case that is missed here is that when the LOC is restarted, all commands issued on the LOC side are lost. I suppose for now this is a small problem that we can live with.

Reply:
While things will work based on how you describe it, calling this a sequence number is misleading and likely confusing (it looks more like a virtual clock if I don't misremember the Lamport paper from the early days of distributed computing).
And having to special case b.2 seems like unneeded extra complexity.
Perhaps there is also a case when both EVE and LOC starts the first time which you need to handle?
In such a case if commands are sent from the LOC with seqno = 0, EVE processes them, sends back seq_ack=1, LOC receives this and updates it seq and sends the same commands again. There is a risk that commands get processed twice in such a case.

I think that if you want to do sequence number semantics you should use something common such as  the TCP SYN exchange where the two ends reliably agree on a pair of sequence numbers and retain state, plus can detect when the peer has restarted and lost state. (There are other reliable transport protocols such as SCTP or QUIC which take a slightly different approach, but my point is inventing a new protocol for sequence number coordinating is likely to lead to new complexity.)

Alternatively use much simpler semantics of a timestamp or random nonce which is echo'ed.
Is there a case where that would be insufficient? Is there a case where the LOC or EVE must reject something which is old which can not be solved with either a random nonce or by a timestamp?

Reply:
The ack echoes back the seq, once processed, as you described. So eve sends ack=0 for the first time in request and if there is something to execute, loc sends seq=1, when eve processes the compound config it sends ack=1 back, which means for the loc "commands below or equal to ack are processed". This is similar to timestamp semantics of the loc, but I use monotonic seq numbers because timestamp can jump into the past, this case I described here: https://github.com/lf-edge/eve-api/pull/27#discussion_r1317123004

Reply:
I don't see why we need to define a new protocol here, particularly since it is more complex than it should be as case b.2 indicates (and yes, others might think that is a trivial tweak but to me it indicates underlying complexity. Sorry I've seen this too often in the IETF where folks specify protocols for their whole carreer - it is a lot harder than one thinks.)

I feel if we so this way we need a protocol specification (akin to [RFC 793](https://www.ietf.org/rfc/rfc9293.pdf)) which specifies how the sequence numbers are coordinated between the two end points.

What happens when the sequence numbers are out of whack and the LOC sends a seq which EVE will ignore? Do we need to define an error? Otherwise that will be a silent failure to act on the command.

I also think that in this approach there should be an explicit SYN flag to synchronize when the LOC has forgotten the last sequence number.

Timestamps have none of this complexity; merely assumes that folks run ntp (and it you don't then clocks will drift and downloads from AWS/Azure will fail.)

Reply:
What's interesting, the doc to the `timestamp` field says the following:
```
   // Timestamp to record when the request to run the command was made.
   // The format of the timestamp is not defined. It can be a Unix timestamp
   // or a different time representation. It is not even required for the timestamp
   // to match the real time or to be in-sync with the device clock.
   // What is required, however, is that two successive but distinct requests made
   // for the same application will have different timestamps attached.
   // This requirement applies even between restarts of the Local profile server.
   // A request made after a restart should not have the same timestamp attached
   // as the previous request made for the same application before the restart.
```
https://github.com/lf-edge/eve-api/blob/26edd90857ba8c8c2122cdc0a16f05a9ef569a77/proto/profile/local_profile.proto#L113

It says ` The format of the timestamp is not defined. It is not even required for the timestamp to match the real time ... ` , and `What is required, however, is that two successive but distinct requests .. will have different timestamps attached`  - so having a monotonic sequence number I still hold these requirements and guarantee that LOC never sends "old" sequence (which is not guaranteed for the timestamp).

Reply:
Anyway, please send me your NAK for the sequence number and confirmation that we use timestamp regardless the described issue and I will quickly update PRs.

Reply:
Sorry for being slow in responding; been a very busy week.

The timestamp/nonce approach works as long as the LPS/LOC only has one outstanding command at any given time - basically needs to repeat that until EVE acks it.

Doing anything fancier would imply a subset of a TCP-like protocol with synchronization of sequence numbers, retransmissions (resend the sequence numbers which has not yet been ack'ed), etc.

Down the road we can explore making sure that these commands are all declarative, meaning that the set of commands in a single compound config can be extended if the previous set has not been ack'ed yet. We can talk about this later.